pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,439 @@
+	})
+
+	return m
+}
+
+// onNodeAdd is the event handler for adding Node.
+func (m *NodeLatencyMonitor) onNodeAdd(obj interface{}) {
+	node := obj.(*corev1.Node)
+	m.latencyStore.addNode(node)
+
+	klog.V(4).InfoS("Node added", "Node", klog.KObj(node))
+}
+
+// onNodeUpdate is the event handler for updating Node.
+func (m *NodeLatencyMonitor) onNodeUpdate(oldObj, newObj interface{}) {
+	node := newObj.(*corev1.Node)
+	m.latencyStore.updateNode(node)
+
+	klog.V(4).InfoS("Node updated", "Node", klog.KObj(node))
+}
+
+// onNodeDelete is the event handler for deleting Node.
+func (m *NodeLatencyMonitor) onNodeDelete(obj interface{}) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+			return
+		}
+		node, ok = deletedState.Obj.(*corev1.Node)
+		if !ok {
+			klog.ErrorS(nil, "DeletedFinalStateUnknown contains non-Node object", "obj", deletedState.Obj)
+			return
+		}
+	}
+
+	m.latencyStore.deleteNode(node)
+}
+
+// onNodeLatencyMonitorAdd is the event handler for adding NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorAdd(obj interface{}) {
+	nlm := obj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor added", "NodeLatencyMonitor", klog.KObj(nlm))
+
+	m.updateLatencyConfig(nlm)
+}
+
+// onNodeLatencyMonitorUpdate is the event handler for updating NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorUpdate(oldObj, newObj interface{}) {
+	oldNLM := oldObj.(*v1alpha1.NodeLatencyMonitor)
+	newNLM := newObj.(*v1alpha1.NodeLatencyMonitor)
+	klog.V(4).InfoS("NodeLatencyMonitor updated", "NodeLatencyMonitor", klog.KObj(newNLM))
+
+	if oldNLM.GetGeneration() == newNLM.GetGeneration() {
+		return
+	}
+
+	m.updateLatencyConfig(newNLM)
+}
+
+// updateLatencyConfig updates the latency config based on the NodeLatencyMonitor CRD.
+func (m *NodeLatencyMonitor) updateLatencyConfig(nlm *v1alpha1.NodeLatencyMonitor) {
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	m.latencyConfigChanged <- struct{}{}
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorDelete(obj interface{}) {
+	m.latencyConfig = &LatencyConfig{Enable: false}
+
+	m.latencyConfigChanged <- struct{}{}
+}
+
+// sendPing sends an ICMP message to the target IP address.
+func (m *NodeLatencyMonitor) sendPing(socket net.PacketConn, addr net.IP) error {
+	var requestType icmp.Type
+
+	ip := &net.IPAddr{IP: addr}
+
+	if addr.To4() == nil {
+		requestType = ipv6.ICMPTypeEchoRequest
+	} else {
+		requestType = ipv4.ICMPTypeEcho
+	}
+
+	timeStart := time.Now()
+	seqID := getICMPSeq()
+	body := &icmp.Echo{
+		ID:   os.Getpid() & 0xffff,                       // Use the current process ID as the ICMP ID
```

Comment:
typically you would want to avoid doing a system call (which involves a context switch) if you don't really need to. So I would suggest calling `os.Getpid()` only once. You could do it in `NewNodeLatencyMonitor` (for example), and store the pid in the struct. Or, you could just define a "global" variable for this file.

Reply:
Yes, once call `Getpid()` is enough, I have modified it.